krabcake/kc_main.c

```diff
@@ -113,14 +113,790 @@ static void kc_post_clo_init(void)
+   // unconditionally evaluated.
+
+   tl_assert(st->tag == Ist_StoreG);
+   IRTypeEnv* tyenv = sbOut->tyenv;
+   IRStoreG* sg       = st->Ist.StoreG.details;
+
+   /* FIXME */
+
+   IRExpr *store_guard = sg->guard;
+   IRExpr *store_addr  = sg->addr;
+   IRExpr *store_data  = sg->data;
+
+   /* FIXME: This unconditionally calls the tracing for both true and false
+      guards, relying on the callee to not inspect the memory if the guard is
+      false.  It would be better to not invoke the tracing routine at all if the
+      guard is false. */
+   store_guard = IRExpr_Unop(Iop_1Uto64, store_guard);
+   store_guard = assignNew('V', sbOut, Ity_I64, store_guard);
+   IRType    store_type = typeOfIRExpr(tyenv, store_data);
+   Int       store_size = sizeofIRType(store_type);
+
+   tl_assert(0); // unimplemented
+   IRDirty* di = unsafeIRDirty_0_N(
+      0,
+      "rs_trace_storeg",
+      VG_(fnptr_to_fnentry)( &rs_trace_storeg ),
+      mkIRExprVec_3(store_guard, store_addr, mkIRExpr_HWord(store_size))
+      );
+   addStmtToIRSB( sbOut, st );
+}
+
+static void kc_instrument_loadg ( IRSB* sbOut,
+                                  IRStmt *st)
+{
+   tl_assert(st->tag == Ist_LoadG);
+   IRTypeEnv* tyenv = sbOut->tyenv;
+   IRLoadG* lg       = st->Ist.LoadG.details;
+   IRType   type     = Ity_INVALID; /* loaded type */
+   IRType   typeWide = Ity_INVALID; /* after implicit widening */
+   typeOfIRLoadGOp(lg->cvt, &typeWide, &type);
+   Int      load_size = sizeofIRType(type);
+   Int      load_widened_size = sizeofIRType(typeWide);
+
+   IRExpr *load_guard = lg->guard;
+   IRExpr *load_addr  = lg->addr;
+
+#if 0
+   VG_(printf)("load_guard pre ");
+   ppIRExpr(load_guard);
+   VG_(printf)(" : ");
+   ppIRType(typeOfIRExpr(tyenv, load_guard));
+   VG_(printf)("\n");
+#endif
+   
+   /* FIXME: This unconditionally calls the tracing for both true and false
+      guards, relying on the callee to not inspect the memory if the guard is
+      false.  It would be better to not invoke the tracing routine at all if the
+      guard is false. */
+   load_guard = IRExpr_Unop(Iop_1Uto64, load_guard);
+   load_guard = assignNew('V', sbOut, Ity_I64, load_guard);
+
+   IRDirty* di = unsafeIRDirty_0_N(
+      0,
+      "rs_trace_loadg",
+      VG_(fnptr_to_fnentry)( &rs_trace_loadg ),
+      mkIRExprVec_4(load_guard, load_addr, mkIRExpr_HWord(load_size), mkIRExpr_HWord(load_widened_size))
+      );
+   addStmtToIRSB( sbOut, IRStmt_Dirty(di) );
+   addStmtToIRSB( sbOut, st );
+}
+
+static void kc_instrument_wrtmp ( IRSB* sbOut,
+                                  IRStmt *st,
+                                  IRExpr* shadow_data )
+{
+   // `t<tmp> = <data>` assigns value to an (SSA) temporary.
+   tl_assert(st->tag == Ist_WrTmp);
+   IRTypeEnv* tyenv = sbOut->tyenv;
+   IRTemp  lhs_tmp = st->Ist.WrTmp.tmp;
+   IRExpr* data = st->Ist.WrTmp.data;
+   IRType  type = typeOfIRExpr(tyenv, data);
+   /* FIXME */
```

Comment:
could you add more details this FIXME?

Reply:
Heh. There's a lot of FIXME's, and unfortunately I didn't record my thoughts on all of them.

I *think* this fixme is about trying to decide whether the rs_trace_wrtmp routine that we call out to here needs to have a copy of the evaluated `data`, or if its sufficient for it to *solely* observe the `lhs_tmp` and the `shadow_data`...

Reply:
anyway, I'll go over the file and see if I can fill in the missing blanks there.